File: content/2013/go4python/monkey.py
#!/usr/bin/python

import urllib

auth_url = 'http://google.com'

def auth(usr):
	try:
		r = urllib.urlopen(auth_url + '/' + usr)
		return r.getcode() == 200
	except:
		return False

def say_hi(usr):
	if auth(usr):
		print 'Hi, %s' % usr
	else:
		print 'unknown user %s' % usr

def sayhitest():
	# Test authenticated user
	globals()['auth'] = lambda x: True
	say_hi('John')

	# Test unauthenticated user
	globals()['auth'] = lambda x: False
	say_hi('John')
#done OMIT

sayhitest()

File: content/2013/go4python/monkey.go
// +build OMIT

package main

import (
	"fmt"
	"net/http"
)

var authURL = ""

var auth = func(user string) bool {
	res, err := http.Get(authURL + "/" + user)
	return err == nil && res.StatusCode == http.StatusOK
}

func sayHi(user string) {
	if !auth(user) {
		fmt.Printf("unknown user %v\n", user)
		return
	}
	fmt.Printf("Hi, %v\n", user)
}

func TestSayHi() {
	auth = func(string) bool { return true }
	sayHi("John")

	auth = func(string) bool { return false }
	sayHi("John")
}

func init() {
	auth = func(string) bool { return true }
}

func TestAnythingElse() {
	// auth has been already set to the fake version
}

func main() {
	TestSayHi()
}
